src/ILCompiler/src/Program.cs

```diff
@@ -268,7 +268,7 @@ private int Run(string[] args)
 
             if (_isWasmCodegen)
                 _targetArchitecture = TargetArchitecture.Wasm32;
-            else if (_isCppCodegen)
+            else if (_isCppCodegen && ((_targetArchitecture == TargetArchitecture.X64) || (_targetArchitecture == TargetArchitecture.ARM64)))
```

Comment:
This makes CppCogeGen on x64&arm64 to work differently from CppCodeGen for other targets. It does not sound right.

I am wondering whether it would be better to fix this by deleting `TargetArchitecture.Cpp64`. Looking at what it is used for, I am not sure we really need it right now. If/once we need it, it may rather be TargetAbi.

@MichalStrehovsky ?

Reply:
> I am wondering whether it would be better to fix this by deleting `TargetArchitecture.Cpp64`

This depends whether we expect x64-Cpp64 to generate different code from arm64-Cpp64 (the output is not actually portable). We fundamentally need to be able to distinguish between 4-byte pointer sizes and 8-byte pointer sizes in CppCodegen because that size is used in several places where piping this through as a `sizeof(void*)` expression wouldn't be possible without a major overhaul (type layout, for example). If we indeed expect differences between various 64bit flavors, deleting Cpp64 is the right course of action.

Cpp64 is currently only used to make the compiler crash and burn if an assembly helper is accidentally included in CppCodegen compilation ([here](https://github.com/dotnet/corert/blob/8df841f1c1235ba161efae95677b62443cb5736d/src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/AssemblyStubNode.cs#L74)). I could live without that protection.

Reply:
> expression wouldn't be possible without a major overhaul (type layout, for example).

That was my thought as well. The type layout rules are different between platforms - e.g. Linux x86 vs. ARM do not have the same field alignment. So as long as we do the type layout the way it is done right now, we need the actual `TargetArchitecture` even for CppCodeGen.